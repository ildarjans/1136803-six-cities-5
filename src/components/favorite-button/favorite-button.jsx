import React, {useState} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {updateFavoriteOffer} from "../../middleware/thunk-api";
import {selectIsAuthStatus} from "../../selectors/selectors";
import {processActionCreator} from "../../store/process/process-action";
import {AppRoute} from "../../const";

const FavoriteButtonComponent = (props) => {
  const {
    updateOfferFavoriteField,
    redirectLoginPage,
    isFavorite,
    options,
    isAuth,
    id,
  } = props;

  const currentClassName = `${options.CLASSNAME_PREFIX}__bookmark-button`;

  const [favoriteState, setFavorite] = useState(isFavorite);

  const handleButtonClick = () => {
    if (!isAuth) {
      redirectLoginPage();
      return;
    }
    updateOfferFavoriteField(id, Number(!favoriteState))
      .then((setFavorite(!favoriteState)));
  };

  return (
    <button
      className={`${currentClassName} ${favoriteState ? `${currentClassName}--active` : ``} button`}
      type="button"
      onClick={handleButtonClick}
    >
      <svg
        className={`${options.CLASSNAME_PREFIX}__bookmark-icon`}
        width={options.WIDTH}
        height={options.WIDTH}
      >
        <use xlinkHref="#icon-bookmark"/>
      </svg>
      <span className="visually-hidden">{favoriteState ? `In` : `To`} bookmarks</span>
    </button>
  );
};

FavoriteButtonComponent.propTypes = ({
  updateOfferFavoriteField: PropTypes.func.isRequired,
  redirectLoginPage: PropTypes.func.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  isAuth: PropTypes.bool.isRequired,
  id: PropTypes.number.isRequired,
  options: PropTypes.shape({
    WIDTH: PropTypes.number.isRequired,
    HEIGHT: PropTypes.number.isRequired,
    CLASSNAME_PREFIX: PropTypes.string.isRequired
  }),
});

const mapDispatchToProps = (dispatch) => ({
  updateOfferFavoriteField(id, status) {
    return dispatch(updateFavoriteOffer(id, status));
  },
  redirectLoginPage() {
    dispatch(processActionCreator.redirectToRoute(AppRoute.LOGIN));
  }
});

const mapStateToProps = (state) => ({
  isAuth: selectIsAuthStatus(state),
});

export const FavoriteButton = connect(mapStateToProps, mapDispatchToProps)(FavoriteButtonComponent);
