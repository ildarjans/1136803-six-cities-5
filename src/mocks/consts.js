export const MOCK_OFFERS_COUNT = 100;
export const MOCK_REVIEWS_COUNT = 5;
export const AVATAR_URL = `https://avatars.dicebear.com/api`;
export const MAX_FEATURES = 3;
export const imageGallery = {
  MIN: 2,
  MAX: 8,
};

export const priceRange = {
  MIN: 19,
  MAX: 4999,
};
export const bedroomsRange = {
  MIN: 1,
  MAX: 6,
};
export const guestRange = {
  MIN: 1,
  MAX: 10,
};

export const avatarSprites = [
  `male`,
  `female`,
  `human`,
  `bottts`,
  `avataaars`,
  `gridy`,
];


export const names = [
  `James`,
  `John`,
  `Robert`,
  `William`,
  `Joseph`,
  `Charles`,
  `Thomas`,
  `Christopher`,
  `Anthony`,
  `Mark`,
  `Paul`,
  `Steven`,
  `Kenneth`,
  `Edward`,
  `Kevin`,
  `Ronald`,
  `Jason`,
  `Jeffrey`,
  `Gary`,
  `Nicholas`,
  `Stephen`,
  `Jonathan`,
  `Frank`,
  `Justin`,
  `Gregory`,
  `Samuel`,
  `Mary`,
  `Patricia`,
  `Elizabeth`,
  `Jessica`,
  `Sarah`,
  `Dorothy`,
  `Karen`,
  `Nancy`,
  `Sandra`,
  `Jennifer`,
  `Amanda`,
  `Rebecca`,
  `Sharon`,
  `Cynthia`,
  `Pamela`,
  `Katherine`,
  `Nicole`,
  `Samantha`,
  `Carolyn`,
];
export const lastNames = [
  `Smith`,
  `Brown`,
  `Johnson`,
  `Jones`,
  `Williams`,
  `Davis`,
  `Clark`,
  `White`,
  `Moore`,
  `Thompson`,
  `Adams`,
  `Thomas`,
  `Anderson`,
  `Lewis`,
  `Jackson`,
  `Green`,
  `Wood`,
  `Parker`,
  `Campbell`,
  `Robinson`,
  `Scott`,
  `Phillips`,
  `Turner`,
  `Carter`,
];

export const features = [
  `Wi-fi`,
  `Dishwasher`,
  `Parking`,
  `Washer`,
  `Elevator`,
  `Conditioner`,
  `Gym`,
  `Guide`,
  `Video games`,
  `Towels`,
  `Kitchen`,
  `Heating`,
];

export const descriptions = [
  `Lorem ipsum dolor sit amet, consectetur adipiscingâ€¦empor incididunt ut labore et dolore magna aliqua`,
  `Odio aenean sed adipiscing diam donec adipiscing tristique risus nec`,
  `Ut ornare lectus sit amet est placerat`,
  `Habitant morbi tristique senectus et`,
  `Ac tortor vitae purus faucibus ornare suspendisse sed nisi lacus`,
  `Tortor aliquam nulla facilisi cras fermentum odio eu feugiat pretium`,
  `Mattis enim ut tellus elementum`,
  `Justo donec enim diam vulputate ut pharetra sit amet aliquam`,
  `Elit ut aliquam purus sit amet luctus venenatis lectus`,
  `Eget velit aliquet sagittis id`,
  `Mi sit amet mauris commodo quis imperdiet massa`,
  `Ullamcorper a lacus vestibulum sed arcu non odio`,
  `Elementum eu facilisis sed odio morbi quis commodo odio`,
  `Et malesuada fames ac turpis egestas integer eget aliquet nibh`,
  `Pharetra massa massa ultricies mi quis hendrerit dolor`,
  `Eu turpis egestas pretium aenean pharetra magna ac`,
  `Dictum fusce ut placerat orci`,
  `Nisi vitae suscipit tellus mauris a diam maecenas sed`,
  `Mauris sit amet massa vitae tortor condimentum lacinia quis vel`,
  `Id faucibus nisl tincidunt eget nullam non nisi est sit.`,
];

export const titles = [
  `Beautiful & luxurious apartment at great location`,
  `Couples particularly like this location `,
  `Big company apartment`,
  `Nearest airport & transport station`,
  `Quite place for zen vacations`,
  `Best lobby bar & Michelin restaurant`,
  `Bed and breakfast. Best offer!`,
];
