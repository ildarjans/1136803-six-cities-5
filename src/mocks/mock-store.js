import configureStore from "redux-mock-store";
import {
  mockCityOffers,
  mockFavoritesOffers,
  mockOffers,
  mockReviews,
  mockUser
} from "./mock-data";

const createStore = configureStore();

const mockStoreTemplate = {
  OFFERS: {
    offers: mockOffers,
    cityOffers: mockCityOffers,
    loading: true,
    error: null
  },
  REVIEWS: {
    reviews: mockReviews,
    loading: false,
    error: null,
    posting: false
  },
  NEAR_OFFERS: {
    nearOffers: mockCityOffers.AMSTERDAM.slice(0, 3),
    loading: false,
    error: null
  },
  PROCESS: {
    activeCity: `AMSTERDAM`,
    hoveredOfferId: ``,
    sortType: ``
  },
  USER: {
    authorizationStatus: `NO_AUTHORIZED`,
    waitingAuthorizationResponse: false,
    error: null,
    user: mockUser
  },
  FAVORITES: {
    favoriteCityOffers: mockFavoritesOffers,
    updating: false,
    loading: false,
    error: null,
  }
};

export const mockStore = createStore(mockStoreTemplate);
