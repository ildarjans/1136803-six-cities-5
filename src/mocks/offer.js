import {
  generateId,
  getLocationCoords,
  getRandomArrayElement,
  getRandomArrayElements,
  getRandomAvatar,
  getRandomFullName,
  getRandomImageGallery,
  getRandomInteger,
} from "./utils";
import {offerTypes, cities} from "../const";
import {
  descriptions,
  features,
  BEDROOMS_RANGE,
  GUEST_RANGE,
  MAX_FEATURES,
  PRICE_RANGE,
  titles,
} from "./consts";

export function getOffer() {
  const city = getRandomArrayElement(cities);
  return {
    city,
    id: generateId(),
    images: getRandomImageGallery(),
    premium: Boolean(getRandomInteger(1)),
    price: getRandomInteger(PRICE_RANGE.MAX, PRICE_RANGE.MIN),
    title: getRandomArrayElement(titles),
    type: getRandomArrayElement(offerTypes),
    rating: getRandomInteger(5),
    description: getRandomArrayElement(descriptions),
    bedrooms: getRandomInteger(BEDROOMS_RANGE.MAX, BEDROOMS_RANGE.MIN),
    guests: getRandomInteger(GUEST_RANGE.MAX, GUEST_RANGE.MIN),
    features: getRandomArrayElements(features, MAX_FEATURES),
    owner: {
      avatar: getRandomAvatar(),
      name: getRandomFullName(),
      super: Boolean(getRandomInteger(1)),
    },
    coords: getLocationCoords(city)
  };
}

