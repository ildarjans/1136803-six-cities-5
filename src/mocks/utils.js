import {
  AVATAR_URL,
  AVATAR_SPRITES,
  LAST_NAMES,
  NAMES,
  IMAGE_GALLERY,
} from "./consts";

export function getRandomInteger(max, min = 0) {
  return min + Math.floor(Math.random() * (max + 1 - min));
}

export function getRandomArrayElements(arr, count) {
  if (arr.length < count) {
    return undefined;
  }
  const resultArray = [];
  while (resultArray.length < count) {
    const randElement = arr[getRandomInteger(arr.length - 1)];
    if (!resultArray.includes(randElement)) {
      resultArray.push(randElement);
    }
  }
  return resultArray;
}

export function getRandomArrayElement(arr) {
  if (!arr && arr.length === 0) {
    return undefined;
  }
  return arr[getRandomInteger(arr.length - 1)];
}

export function getRandomFullName() {
  return (
    `${getRandomArrayElement(NAMES)} ${getRandomArrayElement(LAST_NAMES)}`
  );
}

export function getRandomAvatar() {
  return `${AVATAR_URL}/${getRandomArrayElement(AVATAR_SPRITES)}/${Math.random()}.svg`;
}

export function getRandomImageGallery() {
  const gallerySize = getRandomInteger(IMAGE_GALLERY.MAX, IMAGE_GALLERY.MIN);
  return Array(gallerySize).fill().map(getRandomPhoto);
}

function getRandomPhoto() {
  return `http://picsum.photos/248/152?r=${Math.random()}`;
}

export function generateId() {
  return getRandomInteger(Date.now()).toString(16);
}
