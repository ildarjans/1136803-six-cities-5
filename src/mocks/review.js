import {
  getRandomArrayElement,
  getRandomAvatar,
  getRandomFullName,
  getRandomInteger,
} from "./utils";
import {descriptions} from "./consts";

export function getReview() {
  return {
    avatar: getRandomAvatar(),
    name: getRandomFullName(),
    date: new Date(),
    rating: getRandomInteger(5),
    review: getRandomArrayElement(descriptions),
  };
}

