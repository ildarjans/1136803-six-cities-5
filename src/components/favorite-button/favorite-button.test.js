import React from "react";
import renderer from "react-test-renderer";

import {FAVORITE_BUTTON_OPTIONS} from "../../mocks/mock-data";

import {FavoriteButtonComponent} from "./favorite-button";

const noop = () => {};

describe(`Should favorite button rendered correctly`, () => {
  it(`(1) render Place card implementation if favorite`, () => {
    const tree = renderer
      .create(
          <FavoriteButtonComponent
            options={FAVORITE_BUTTON_OPTIONS.PLACE_CARD}
            updateOfferFavoriteField={noop}
            redirectLoginPage={noop}
            isFavorite={true}
            isAuth={true}
            id={1}
          />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it(`(2) render Place card implementation if not favorite`, () => {
    const tree = renderer
      .create(
          <FavoriteButtonComponent
            options={FAVORITE_BUTTON_OPTIONS.PLACE_CARD}
            updateOfferFavoriteField={noop}
            redirectLoginPage={noop}
            isFavorite={false}
            isAuth={false}
            id={1}
          />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it(`(3) render Property page implementation if favorite`, () => {
    const tree = renderer
      .create(
          <FavoriteButtonComponent
            options={FAVORITE_BUTTON_OPTIONS.PROPERTY}
            updateOfferFavoriteField={noop}
            redirectLoginPage={noop}
            isFavorite={true}
            isAuth={true}
            id={1}
          />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it(`(4) render Property page implementation if not favorite`, () => {
    const tree = renderer
      .create(
          <FavoriteButtonComponent
            options={FAVORITE_BUTTON_OPTIONS.PROPERTY}
            updateOfferFavoriteField={noop}
            redirectLoginPage={noop}
            isFavorite={false}
            isAuth={false}
            id={1}
          />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });

});
