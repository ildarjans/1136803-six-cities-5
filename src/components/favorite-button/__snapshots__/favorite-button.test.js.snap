// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should favorite button rendered correctly (1) render Place card implementation if favorite 1`] = `
<button
  className="place-card__bookmark-button place-card__bookmark-button--active button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={18}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    In
     bookmarks
  </span>
</button>
`;

exports[`Should favorite button rendered correctly (2) render Place card implementation if not favorite 1`] = `
<button
  className="place-card__bookmark-button  button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={18}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To
     bookmarks
  </span>
</button>
`;

exports[`Should favorite button rendered correctly (3) render Property page implementation if favorite 1`] = `
<button
  className="property__bookmark-button property__bookmark-button--active button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="property__bookmark-icon"
    height={31}
    width={31}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    In
     bookmarks
  </span>
</button>
`;

exports[`Should favorite button rendered correctly (4) render Property page implementation if not favorite 1`] = `
<button
  className="property__bookmark-button  button"
  onClick={[Function]}
  type="button"
>
  <svg
    className="property__bookmark-icon"
    height={31}
    width={31}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To
     bookmarks
  </span>
</button>
`;
